File: crates/loader/tests/hmr_lifecycle_oracle.rs
//! Source/native Host reload publication, provider visibility, and lifecycle ordering.

use std::{
    fs::OpenOptions,
    io::Write as _,
    path::{Path, PathBuf},
    sync::{
        Arc,
        atomic::{AtomicBool, Ordering},
    },
};

use parking_lot::Mutex;
use seekdeep_cordis::{Context, EventOptions, EventReply, FiberState, PluginFiber};
use seekdeep_loader::{LOADER, PluginCatalog};
use serde_json::{Value, json};

fn record(path: &Path, value: &Value) -> anyhow::Result<()> {
    let mut encoded = serde_json::to_vec(value)?;
    encoded.push(b'\n');
    OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)?
        .write_all(&encoded)?;
    Ok(())
}

fn trace(path: &Path) -> anyhow::Result<Value> {
    std::fs::read_to_string(path)?
        .lines()
        .map(serde_json::from_str)
        .collect::<Result<Vec<_>, _>>()
        .map(Value::Array)
        .map_err(Into::into)
}

const fn state(state: FiberState) -> u8 {
    match state {
        FiberState::Pending => 0,
        FiberState::Loading => 1,
        FiberState::Active => 2,
        FiberState::Failed => 3,
        FiberState::Disposed => 4,
        FiberState::Unloading => 5,
    }
}

fn plugin(generation: &str, failure: &str, async_dispose: bool) -> String {
    format!(
        "import {{ appendFileSync }} from 'node:fs';\n\
         const record = kind => appendFileSync(new URL('trace.jsonl', import.meta.url), JSON.stringify({{kind}}) + '\\n');\n\
         export const name = {generation:?};\n\
         export const generation = {generation:?};\n\
         export function apply(ctx, config) {{\n\
           record('apply:' + generation + ':' + config.id);\n\
           ctx.provide('probe_' + config.id, generation);\n\
           ctx.effect(() => {asynchronous}() => {{\n\
             record('dispose-start:' + generation + ':' + config.id);\n\
             {wait}\n\
             record('dispose-end:' + generation + ':' + config.id);\n\
           }});\n\
           if ({failure:?} === 'body' && config.id === 'second') throw new Error('second replacement apply rejected');\n\
         }}\napply.generation = generation;\n",
        asynchronous = if async_dispose { "async " } else { "" },
        wait = if async_dispose {
            "await Promise.resolve();"
        } else {
            ""
        },
    )
}

#[derive(Clone)]
struct Observation {
    context: Context,
    trace: PathBuf,
    first_uid: Arc<Mutex<Option<u64>>>,
    old: Arc<Mutex<Vec<Arc<PluginFiber>>>>,
    reject: Arc<AtomicBool>,
}

impl Observation {
    fn uid(&self, fiber: &PluginFiber) -> Option<u64> {
        fiber.uid().map(|uid| uid - self.first_uid.lock().unwrap())
    }

    fn fiber(&self, fiber: &PluginFiber) -> Value {
        json!({"id":fiber.entry_id(), "uid":self.uid(fiber), "state":state(fiber.fiber().state())})
    }

    fn provider(&self, id: &str) -> Value {
        let name = format!("probe_{id}");
        json!({"id":id,
            "strict":self.context.get_named::<Value>(&name).as_deref(),
            "loose":self.context.get_named_relaxed::<Value>(&name).as_deref()})
    }

    fn publication(&self, fiber: &PluginFiber) -> anyhow::Result<()> {
        let config = fiber.config();
        let Some(id @ ("first" | "second")) = config["id"].as_str() else {
            return Ok(());
        };
        {
            let mut first_uid = self.first_uid.lock();
            if first_uid.is_none() {
                *first_uid = fiber.uid();
            }
        }
        let name = format!("probe_{id}");
        record(
            &self.trace,
            &json!({"kind":"publication", "id":id, "generation":fiber.plugin_name(),
            "uid":self.uid(fiber), "state":state(fiber.fiber().state()),
            "strict":self.context.get_named::<Value>(&name).as_deref(),
            "loose":self.context.get_named_relaxed::<Value>(&name).as_deref()}),
        )?;
        if fiber.uid().is_some() && id == "second" && self.reject.swap(false, Ordering::AcqRel) {
            record(&self.trace, &json!({"kind":"registration-failure:second"}))?;
            anyhow::bail!("second replacement registration rejected");
        }
        Ok(())
    }

    fn reload(&self) -> anyhow::Result<()> {
        let entries = self
            .context
            .registry()
            .values()
            .into_iter()
            .flat_map(|runtime| runtime.fibers)
            .filter(|fiber| matches!(fiber.entry_id().as_deref(), Some("first" | "second")))
            .map(|fiber| self.fiber(&fiber))
            .collect::<Vec<_>>();
        record(
            &self.trace,
            &json!({"kind":"event:hmr/reload",
            "oldFibers":self.old.lock().iter().map(|fiber|self.fiber(fiber)).collect::<Vec<_>>(),
            "entries":entries,
            "providers":[self.provider("first"),self.provider("second")]}),
        )
    }

    fn install(&self) -> anyhow::Result<()> {
        let observation = self.clone();
        self.context.events().on(
            &self.context,
            "internal/plugin",
            move |_, args| {
                let fiber = args.get::<PluginFiber>(0).unwrap();
                let result = observation.publication(&fiber);
                Box::pin(async move {
                    result?;
                    Ok(EventReply::Undefined)
                })
            },
            EventOptions::default(),
        )?;
        let observation = self.clone();
        self.context.events().on(
            &self.context,
            "hmr/reload",
            move |_, _| {
                let result = observation.reload();
                Box::pin(async move {
                    result?;
                    Ok(EventReply::Undefined)
                })
            },
            EventOptions::default(),
        )?;
        Ok(())
    }
}

async fn scenario(failure: &str, async_dispose: bool) -> anyhow::Result<Value> {
    let temporary = tempfile::tempdir()?;
    let filename = temporary.path().join("plugin.mjs");
    let tracefile = temporary.path().join("trace.jsonl");
    let context = Context::new();
    let observation = Observation {
        context: context.clone(),
        trace: tracefile.clone(),
        first_uid: Arc::default(),
        old: Arc::default(),
        reject: Arc::new(AtomicBool::new(false)),
    };
    observation.install()?;
    std::fs::write(&filename, plugin("old", "none", async_dispose))?;
    let composition = PluginCatalog::new().load_yaml_at(
        &context,
        "- id: first\n  name: ./plugin.mjs\n  config: { id: first }\n- id: second\n  name: ./plugin.mjs\n  config: { id: second }\n",
        temporary.path().join("cordis.yml"),
    ).await?;
    *observation.old.lock() = composition.fibers();
    let initial = trace(&tracefile)?;
    std::fs::write(&tracefile, "")?;
    std::fs::write(&filename, plugin("new", failure, async_dispose))?;
    observation
        .reject
        .store(failure == "registration", Ordering::Release);
    let replaced = composition.reload_module(&filename).await;
    if failure == "registration" {
        assert!(
            replaced
                .unwrap_err()
                .to_string()
                .contains("second replacement registration rejected")
        );
    } else {
        replaced?;
    }
    let settlement_error = context
        .get(LOADER)
        .unwrap()
        .wait()
        .await
        .err()
        .map(|error| error.to_string());
    let old = observation.old.lock().clone();
    PluginFiber::await_all_quiescent(&old).await;
    let settled = trace(&tracefile)?;
    let snapshots = composition.entries();
    let entries = composition
        .fibers()
        .into_iter()
        .map(|fiber| {
            let id = fiber.entry_id().unwrap();
            let snapshot = snapshots
                .iter()
                .find(|entry| entry.id.as_str() == id)
                .unwrap();
            json!({"id":id,"uid":observation.uid(&fiber),"state":state(fiber.fiber().state()),
            "disabled":snapshot.disabled,"generation":fiber.plugin_name()})
        })
        .collect::<Vec<_>>();
    composition.dispose().await?;
    Ok(
        json!({"failure":failure,"asyncDispose":async_dispose,"initial":initial,"settled":settled,
        "entries":entries,"settlementError":settlement_error}),
    )
}

fn source_reports() -> anyhow::Result<Value> {
    let root = Path::new(env!("CARGO_MANIFEST_DIR"));
    let output = std::process::Command::new("node")
        .args(["--experimental-transform-types", "--expose-internals"])
        .arg(root.join("tests/fixtures/hmr_lifecycle_oracle.mjs"))
        .arg(root.join("../../../deepseek-harness"))
        .output()?;
    anyhow::ensure!(
        output.status.success(),
        "{}",
        String::from_utf8_lossy(&output.stderr)
    );
    Ok(serde_json::from_slice(&output.stdout)?)
}

#[tokio::test]
async fn real_node_reload_matches_all_six_source_lifecycle_traces() -> anyhow::Result<()> {
    let expected = source_reports()?;
    let mut actual = Vec::new();
    for async_dispose in [false, true] {
        for failure in ["none", "body", "registration"] {
            actual.push(
                tokio::time::timeout(
                    std::time::Duration::from_secs(10),
                    scenario(failure, async_dispose),
                )
                .await??,
            );
        }
    }
    let actual = Value::Array(actual);
    if let Some(directory) = std::env::var_os("SEEKDEEP_HMR_TRACE_DIR") {
        let directory = PathBuf::from(directory);
        std::fs::create_dir_all(&directory)?;
        std::fs::write(
            directory.join("source-hmr-lifecycle.json"),
            serde_json::to_vec_pretty(&expected)?,
        )?;
        std::fs::write(
            directory.join("native-hmr-lifecycle.json"),
            serde_json::to_vec_pretty(&actual)?,
        )?;
    }
    assert_eq!(actual, expected);
    Ok(())
}

File: crates/loader/tests/fixtures/hmr_lifecycle_oracle.mjs
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { registerHooks } from 'node:module';
import { pathToFileURL } from 'node:url';

const source = process.argv[2];
registerHooks({
  resolve(specifier, context, nextResolve) {
    const entry = {
      '@deepseek-ai/cordis': 'vendor/cordis/src/index.ts',
      '@deepseek-ai/cordis-plugin-loader': 'vendor/loader/src/index.ts',
    }[specifier];
    return entry
      ? { url: pathToFileURL(path.join(source, entry)).href, shortCircuit: true }
      : nextResolve(specifier, context);
  },
});
const { Context } = await import(pathToFileURL(path.join(source, 'vendor/cordis/src/index.ts')));
const { Loader } = await import(pathToFileURL(path.join(source, 'vendor/loader/src/index.ts')));
const { TimerService } = await import(pathToFileURL(path.join(source, 'vendor/timer/src/index.ts')));
const { default: Hmr } = await import(pathToFileURL(path.join(source, 'vendor/hmr/src/index.ts')));

function plugin(generation, failure, asyncDispose) {
  return `
import { appendFileSync } from 'node:fs';
const record = kind => appendFileSync(new URL('trace.jsonl', import.meta.url), JSON.stringify({kind}) + '\\n');
export const name = ${JSON.stringify(generation)};
export const generation = ${JSON.stringify(generation)};
export function apply(ctx, config) {
  record('apply:' + generation + ':' + config.id);
  ctx.provide('probe_' + config.id, generation);
  ctx.effect(() => ${asyncDispose ? 'async ' : ''}() => {
    record('dispose-start:' + generation + ':' + config.id);
    ${asyncDispose ? 'await Promise.resolve();' : ''}
    record('dispose-end:' + generation + ':' + config.id);
  });
  if (${JSON.stringify(failure)} === 'body' && config.id === 'second') throw new Error('second replacement apply rejected');
}
apply.generation = generation;
`;
}

async function scenario(failure, asyncDispose) {
  const root = fs.realpathSync(fs.mkdtempSync(path.join(os.tmpdir(), 'seekdeep-source-hmr-lifecycle-')));
  const filename = path.join(root, 'plugin.mjs');
  const tracefile = path.join(root, 'trace.jsonl');
  const record = value => fs.appendFileSync(tracefile, JSON.stringify(value) + '\n');
  const trace = () => fs.readFileSync(tracefile, 'utf8').split('\n').filter(Boolean).map(JSON.parse);
  const baseUrl = pathToFileURL(root + path.sep).href;
  const ctx = new Context().extend({ baseUrl });
  try {
    await ctx.plugin(Loader, { baseUrl }).await();
    ctx.loader.internal.version = typeof ctx.loader.internal.resolve === 'function' ? 'v1' : 'v2';
    await ctx.plugin(TimerService).await();
    await ctx.plugin(Hmr, { root: [], ignored: [], debounce: 0 }).await();
    let rejectRegistration = false;
    let firstUid;
    ctx.on('internal/plugin', fiber => {
      const id = fiber._config?.id;
      if (!['first', 'second'].includes(id)) return;
      firstUid ??= fiber.uid;
      record({ kind:'publication', id, generation:fiber.runtime.callback.generation, uid:fiber.uid === null ? null : fiber.uid - firstUid, state:fiber.state,
        strict:ctx.get('probe_' + id, true) ?? null, loose:ctx.get('probe_' + id, false) ?? null });
      if (fiber.uid && id === 'second' && rejectRegistration) {
        rejectRegistration = false;
        record({kind:'registration-failure:second'});
        throw new Error('second replacement registration rejected');
      }
    }, { global: true });
    ctx.on('hmr/reload', reloads => {
      record({ kind:'event:hmr/reload',
        oldFibers:[...reloads].flatMap(([, reload]) => [...reload.runtime.fibers].map(fiber => ({id:fiber.entry.id, uid:fiber.uid === null ? null : fiber.uid - firstUid, state:fiber.state}))),
        entries:[...ctx.loader.entries()].map(entry => ({id:entry.id, uid:entry.fiber.uid === null ? null : entry.fiber.uid - firstUid, state:entry.fiber.state})),
        providers:['first','second'].map(id => ({id, strict:ctx.get('probe_' + id, true) ?? null, loose:ctx.get('probe_' + id, false) ?? null})),
      });
    });
    fs.writeFileSync(filename, plugin('old', 'none', asyncDispose));
    await ctx.loader.root.update([
      {id:'first',name:'./plugin.mjs',config:{id:'first'}},
      {id:'second',name:'./plugin.mjs',config:{id:'second'}},
    ]);
    await ctx.loader.await();
    const initial = trace();
    fs.writeFileSync(tracefile, '');
    fs.writeFileSync(filename, plugin('new', failure, asyncDispose));
    rejectRegistration = failure === 'registration';
    ctx.hmr.stashed.add(pathToFileURL(filename).href);
    await ctx.hmr.partialReload();
    let settlementError = null;
    try { await ctx.loader.await(); } catch (error) { settlementError = error.message; }
    const settled = trace();
    const entries = [...ctx.loader.entries()].map(entry => ({
      id:entry.id, uid:entry.fiber.uid === null ? null : entry.fiber.uid - firstUid, state:entry.fiber.state, disabled:entry.options.disabled ?? false,
      generation:entry.fiber.runtime.callback.generation,
    }));
    return {failure, asyncDispose, initial, settled, entries, settlementError};
  } finally {
    if (ctx.loader) await ctx.loader.root.stop();
    await ctx.fiber.dispose();
    fs.rmSync(root, {recursive:true,force:true});
  }
}

const reports = [];
for (const asyncDispose of [false, true]) {
  for (const failure of ['none', 'body', 'registration']) reports.push(await scenario(failure, asyncDispose));
}
process.stdout.write(JSON.stringify(reports) + '\n');
